timemachine/fe/rest/single_topology.py

```diff
@@ -79,37 +81,89 @@ def __init__(
             max_temperature_scale, temperature_scale_interpolation
         )
 
+    @cached_property
+    def rest_region_atom_idxs(self) -> set[int]:
+        """Returns the set of indices of atoms in the combined ligand that are in the REST region."""
+
+        def params_eq(params_a, params_b) -> bool:
+            eq_p = params_a == params_b
+            if isinstance(eq_p, bool):
+                return eq_p
+            elif isinstance(eq_p, (np.ndarray, jax.Array)):
+                return bool(np.all(eq_p))
+            else:
+                assert False
+
+        # Heuristic: include in the rest region atoms involved in bond, angle, or improper torsion interactions that
+        # differ in the end states. Note that proper torsions are omitted from the heuristic as this tends to result in
+        # larger REST regions than seem desirable.
+        aligned_potentials: list[AlignedPotential] = [
+            self.aligned_bond,
+            self.aligned_angle,
+            self.aligned_improper,
+        ]
+
+        idxs = {
+            int(idx)
+            for aligned in aligned_potentials
+            for idxs, params_a, params_b in zip(aligned.idxs, aligned.src_params, aligned.dst_params)
+            if not params_eq(params_a, params_b)
+            for idx in idxs  # type: ignore[attr-defined]
+        }
+
+        # Ensure all dummy atoms are included in the REST region
+        idxs |= self.get_dummy_atoms_a()
+        idxs |= self.get_dummy_atoms_b()
+
+        return idxs
+
     @cached_property
     def aliphatic_ring_bonds(self) -> set[CanonicalBond]:
+        """Returns the set of aliphatic ring bonds in the combined ligand."""
         ring_bonds_a = {bond.translate(self.a_to_c) for bond in get_aliphatic_ring_bonds(self.mol_a)}
         ring_bonds_b = {bond.translate(self.b_to_c) for bond in get_aliphatic_ring_bonds(self.mol_b)}
         ring_bonds_c = ring_bonds_a | ring_bonds_b
         return ring_bonds_c
 
     @cached_property
     def rotatable_bonds(self) -> set[CanonicalBond]:
+        """Returns the set of rotatable bonds in the combined ligand."""
         rotatable_bonds_a = {bond.translate(self.a_to_c) for bond in get_rotatable_bonds(self.mol_a)}
         rotatable_bonds_b = {bond.translate(self.b_to_c) for bond in get_rotatable_bonds(self.mol_b)}
         rotatable_bonds_c = rotatable_bonds_a | rotatable_bonds_b
         return rotatable_bonds_c
 
     @cached_property
     def propers(self) -> list[CanonicalProper]:
+        """Returns a list of proper torsions in the combined ligand."""
         # TODO: refactor SingleTopology to compute src and dst alignment at initialization
         return [mkproper(*idxs) for idxs in super().setup_intermediate_state(0.0).proper.potential.idxs]
 
     @cached_property
-    def target_proper_idxs(self) -> list[int]:
-        return [
-            idx
+    def candidate_propers(self) -> set[tuple[int, CanonicalProper]]:
```

Comment:
Changed the naming slightly so that `candidate_propers` refer to interactions that could potentially be softened if they involve an atom in the REST region, while `target_propers` refer to `candidate_propers` that will be softened because they involve an atom in the REST region. The previous behavior was effectively set `target_propers = candidate_propers`.